plugins/woocommerce-blocks/packages/components/checkbox-control/validated-checkbox-control.tsx

```
@@ -0,0 +1,246 @@
+			( select ) => {
+				const store = select( VALIDATION_STORE_KEY );
+				return {
+					validationError: store.getValidationError( errorIdString ),
+					validationErrorId:
+						store.getValidationErrorId( errorIdString ),
+				};
+			}
+		);
+
+		const validateInput = useCallback(
+			( errorsHidden = true ) => {
+				const inputObject = inputRef.current || null;
+
+				if ( inputObject === null ) {
+					return;
+				}
+
+				if (
+					inputObject.checkValidity() &&
+					customValidationRef.current( inputObject )
+				) {
+					clearValidationError( errorIdString );
+					return;
+				}
+
+				setValidationErrors( {
+					[ errorIdString ]: {
+						message: getValidityMessageForInput(
+							label,
+							inputObject,
+							customValidityMessage
+						),
+						hidden: errorsHidden,
+					},
+				} );
+			},
+			[
+				clearValidationError,
+				errorIdString,
+				setValidationErrors,
+				label,
+				customValidityMessage,
+			]
+		);
+
+		// Allows parent to trigger revalidation.
+		useImperativeHandle(
+			forwardedRef,
+			function () {
+				return {
+					focus() {
+						inputRef.current?.focus();
+					},
+					revalidate() {
+						validateInput( false );
+					},
+				};
+			},
+			[ validateInput ]
+		);
+
+		/**
+		 * Validation on mount.
+		 */
+		useEffect( () => {
+			if ( validateOnMount ) {
+				validateInput( true );
+			}
+		}, [ validateOnMount, validateInput ] );
+
+		// Remove validation errors when unmounted.
+		useEffect( () => {
+			return () => {
+				clearValidationError( errorIdString );
+			};
+		}, [ clearValidationError, errorIdString ] );
+
+		if ( passedErrorMessage !== '' && isObject( validationError ) ) {
+			validationError.message = passedErrorMessage;
+		}
+
+		const hasError = validationError?.message && ! validationError?.hidden;
+
+		return (
+			<CheckboxControl
+				className={ clsx( className, {
+					'has-error': hasError,
+				} ) }
+				aria-invalid={ hasError === true }
```

Not saying this is wrong, but it feels odd. is there a situation where would we have validationError be hidden and mark this as valid? not sure if that even makes sense but it's at least a smell to me. What does validation error being hidden mean when there is a validation error?

---

> What does validation error being hidden mean when there is a validation error?

All required fields have errors added to the validation data state, they start off as hidden, and they get removed when the field becomes valid. The errors become unhidden when the form is submitted.